CSHARP-1847: Add Ubuntu 18.04 as OS in evg config

durran opened the pull request:
https://spruce.mongodb.com/version/5fda1ef93e8e863b554dbd69/tasks

durran committed:
CSHARP-1847: Add Ubuntu 18.04 as OS in evg config

durran commented on CSharpDriver.sln:
```diff
@@ -39,8 +39,6 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "MongoDB.Driver.Legacy.TestH
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "MongoDB.Driver.Legacy.Tests", "tests\MongoDB.Driver.Legacy.Tests\MongoDB.Driver.Legacy.Tests.csproj", "{7A015030-6329-4FAD-B6E3-CED5ED53019C}"
 EndProject
-Project("{F184B08F-C81C-45F6-A57F-5ABD9991F28F}") = "MongoDB.Driver.Legacy.VB.Tests", "tests\MongoDB.Driver.Legacy.VB.Tests\MongoDB.Driver.Legacy.VB.Tests.vbproj", "{D197D685-5509-491E-A456-BA36C662CBB4}"
```
Removed from the solution as it's only testing the legacy driver and was having issues with finding VB on Linux. Code is still there.

JamesKovacs replied:
Do we want to remove the VB.NET tests on Windows too? If we decide the VB.NET tests are no longer valuable, then we should remove the entire project rather than simply removing it from the solution.

durran replied:
At the project level I couldn't find a nice way to only remove for Linux, so we discussed in a a standup and said this was a good temporary solution to get the ball moving forward.

durran commented on Tools/packages.config:
```diff
@@ -1,4 +1,4 @@
 <?xml version="1.0" encoding="utf-8"?>
 <packages>
-    <package id="Cake" version="0.30.0" />
+    <package id="Cake" version="0.38.4" />
```
Moved Cake to latest to fix open issues on previous versions on Linux.

durran commented on build-posix.sh:
```diff
@@ -0,0 +1,8 @@
+dotnet new tool-manifest --force
+dotnet tool install Cake.Tool
+dotnet cake --bootstrap
+dotnet cake "$@"
```
Using Cake.Tool to actually run the build. The Windows script will change to this as well.

durran commented on build.cake:
```diff
@@ -1,8 +1,9 @@
-#addin "nuget:?package=Cake.Incubator&version=3.1.0"
-#tool "nuget:?package=GitVersion.CommandLine&version=4.0.0"
-#tool "nuget:?package=xunit.runner.console"
+#module nuget:?package=Cake.DotNetTool.Module&version=0.4.0
```
This allows Cake to understand the `dotnet` directive and use the tool.

durran commented on build.cake:
```diff
@@ -73,7 +74,7 @@ Task("BuildArtifacts")
     .Does(() =>
     {
-        foreach (var targetFramework in new[] { "net452", "netstandard1.5", "netstandard2.0" })
+        foreach (var targetFramework in new[] { "net452", "netstandard1.5", "netstandard2.0", "netstandard2.1" })
```
Should be testing on .NET Core 3.x as well.

rstam replied:
That should be a change to the "Test" task, not the "BuildArtifacts" task.

I would remove `netstandard2.1` from here because we do NOT target that framework.

durran replied:
Done

durran replied:
I actually came back and removed the support for .NET Core 3 just to keep the PR in line with what we currently do.

durran commented on build.cake:
```diff
@@ -177,6 +180,42 @@ Task("TestAwsAuthentication")
             );
         });
 
+Task("TestPlainAuthentication")
```
These are adding targets for GSSAPI tests and LDAP tests.

rstam replied:
Were these never being tested before? Or were these tests being run a different way?

durran replied:
They were not being run before as part of a cake target.

rstam replied:
Shouldn't running these tests be part of some other ticket?

This doesn't seem related to running on Linux.

durran commented on src/MongoDB.Bson/MongoDB.Bson.csproj:
```diff
@@ -33,6 +33,10 @@
     <DefineConstants>TRACE</DefineConstants>
   </PropertyGroup>
 
+  <ItemGroup>
```
This is required on Linux to pass compilation of the build.cake for when .NET framework is not present or Mono is.

rstam replied:
I don't understand why but I'll trust you on this one.

DmitryLukyanov replied:
Is it related only to cake? I see that compilation pass without this change via `dotnet build`

durran replied:
Yes was directly related to Cake and it's dependencies as there was a direct dependency on .NET Framework in one of the modules.

durran commented on build.ps1:
```diff
@@ -262,26 +234,24 @@ if (!(Test-Path $CAKE_EXE)) {
 Write-Host "Running build script..."
-echo $CAKE_EXE $cakeArguments
-&$CAKE_EXE $cakeArguments
+Invoke-Expression "& $CAKE_EXE_INVOCATION --bootstrap"
```
When using the #tool directive in build.cake, Cake needs to run twice with a bootstrap first.

durran commented on tests/MongoDB.Bson.Tests/Jira/CSharp275Tests.cs:
```diff
@@ -83,35 +83,37 @@ public Test(string json, string iso)
-            new Test("Mon, 10 Oct 2011 11:22:33 W", "2011-10-10T11:22:33+10:00"),
-            new Test("Mon, 10 Oct 2011 11:22:33 X", "2011-10-10T11:22:33+11:00"),
-            new Test("Mon, 10 Oct 2011 11:22:33 Y", "2011-10-10T11:22:33+12:00"),
+#if Windows
```
Failing on Linux with Military time.

rstam replied:
Do we need a JIRA ticket to fix this?

1. Why does it only fail on Linux?
2. If "-1:00" is wrong and "+1:00" is right we have a bug

DmitryLukyanov replied:
`#if Windows` this actually doesn't work as far as I can see. It will always return false even on windows. This `ifdef` variables work in pair with csproj changes, see here: https://stackoverflow.com/questions/30153797/c-sharp-preprocessor-differentiate-between-operating-systems

DmitryLukyanov replied:
Same for all other places

DmitryLukyanov replied:
>Do we need a JIRA ticket to fix this?

agreed

rstam replied:
Notes: preprocessor symbols are usually all uppercase:

    #if WINDOWS

Also, I can't find anywhere that this preprocessor symbol is ever defined, so these tests would now be #ifdef'ed out of existence for ALL platforms.

They also won't be compiled when running tests in Visual Studio.

JamesKovacs replied:
Based on https://militarybenefits.info/military-time/ (found via  quick Google search), Alpha is GMT+1, Bravo is GMT+2, ..., Mike is GMT+12, November is GMT-1, Oscar is GMT-2, etc. We should file a bug to investigate further.

durran replied:
https://jira.mongodb.org/browse/CSHARP-3268

durran commented on tests/MongoDB.Driver.Core.Tests/Core/Compression/SnappyNativeTests.cs:
```diff
@@ -39,6 +39,7 @@ namespace MongoDB.Driver.Core.Tests.Core.Compression
 {
     public class SnappyNativeTests
     {
+#if Windows
```
Snappy not working on Linux yet.

durran commented on tests/MongoDB.Driver.Core.Tests/Core/Compression/ZstandardNativeTests.cs:
```diff
@@ -28,7 +28,8 @@ namespace MongoDB.Driver.Core.Tests.Core.Compression
     public class ZstandardNativeTests
     {
-        #region static
+#if Windows
```
Zstd not working on Linux yet.

durran commented on tests/MongoDB.Driver.Core.Tests/Core/WireProtocol/CommandWriteProtocolTests.cs:
```diff
@@ -91,7 +91,8 @@ public void Execute_should_use_cached_IWireProtocol_if_available([Values(false,
             {
                 var e = exception.Should().BeOfType<ArgumentException>().Subject;
-                e.Message.Should().Be("CommandResponseHandling must be Return, NoneExpected or ExhaustAllowed.\r\nParameter name: responseHandling");
+                e.Message.Should().Contain("CommandResponseHandling must be Return, NoneExpected or ExhaustAllowed.");
```
Line breaks not the same on both platforms so split into 2 assertions.

DmitryLukyanov replied:
Then, we probably need to create a helper that will provide these values regardless of the OS. Definitely, it's not the scope of this PR

durran commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -92,9 +92,10 @@ public void RunTestDefinition(JsonDrivenTestCase testCase)
                 }
             }
-
+#if Windows
```
This and other test failing not finding libmongocrypt

DmitryLukyanov replied:
We should create a ticket to restore all these `ifdefs` conditions

durran replied:
I've already fixed them in this PR.

durran commented:
Open question. I used compile time directives to skip tests on Linux as we had a similar pattern with .NET standard version branches. I do feel, however, that maybe creating a custom attribute is better in most cases, like `SkippableIfNotWindows` or the like. Thoughts? Or maybe make it a separate ticket?

DmitryLukyanov commented:
> 
> 
> Open question. I used compile time directives to skip tests on Linux as we had a similar pattern with .NET standard version branches. I do feel, however, that maybe creating a custom attribute is better in most cases, like `SkippableIfNotWindows` or the like. Thoughts? Or maybe make it a separate ticket?

I would say the current way is more flexible since we can use it in the code (not only in the tests) in future

durran commented on tests/MongoDB.Driver.Tests/Linq/Translators/FindProjectionTranslatorTests.cs:
```diff
@@ -55,8 +55,10 @@ public void Should_not_translate_when_identity_is_present()
+            var projection = result.Projection;
 
-            result.Projection.Should().Be("{ _id: 1, A: 1 }");
+            projection.GetValue("_id").Should().Be(1);
```
Projection order was different on Linux, but values were correct.

jyemin replied:
Odd.  Any hypothesis as to why there's a difference?

DmitryLukyanov replied:
we already had a similar issue here https://github.com/mongodb/mongo-csharp-driver/blob/master/tests/MongoDB.Driver.Legacy.Tests/Builders/UpdateBuilderTests.cs#L842 I think we even have a ticket for this. But I don't remember whether we investigated this or no

rstam replied:
@DmitryLukyanov I think the test you linked to involves the *server* returning fields in a different order.

The part of this test that is failing is executing client side only.

So I don't think they are not the same issue.

I would not have expected this projection to be different on Linux.

DmitryLukyanov replied:
@rstam, I don't remember all details, but I think the server response should not be related to the target framework which is used in the client app, shouldn't it?

DmitryLukyanov replied:
in fact, I think the server doesn't know anything about client TF or at least doesn't consider any logic for this, but maybe I'm missing something

rstam replied:
> but I think the server response should not be related to the target framework which is used in the client app

I believe that to be a true statement.

But that's not what's happening in this test. The Projection being asserted against is a client side result of translating a C# Expression to an MQL projection.

DmitryLukyanov replied:
Ok, I don't insist since I didn't investigate it.

DmitryLukyanov replied:
let me know if I need to make further investigation here

rstam replied:
We still don't understand why this change is necessary.

Maybe there is a hashtable involved somewhere deep inside the LINQ2 implementation that is returning entries in a different order on different platforms (so this only passed by coincidence before now)?

DmitryLukyanov replied:
>@DmitryLukyanov I think the test you linked to involves the server returning fields in a different order.

the mentioned legacy test is related to client-side behavior, the only reason why it fails without running a server is that we initialize a test collection in the constructor for each of the tests in this file, but this initialization is unrelated to the mentioned tests.
I don't say that this is the same issue, but I think they could have the same underlying reason

rstam replied:
Initializing a test collection should have no effect on the `Projection` which is a solely client side operation.

If the fields in the projection are really coming back out of order I think it is a driver bug.

DmitryLukyanov replied:
>Initializing a test collection should have no effect on the Projection which is a solely client side operation.

exactly what I wrote above :)

durran replied:
https://jira.mongodb.org/browse/CSHARP-3284

rstam commented:
Minor note: we have the custom of naming our branches like this: `csharp1847`

It makes life just that litte bit easier if we all name our branches the same way.

rstam commented on build.cake:
```diff
@@ -49,7 +51,11 @@ Task("Release")
-        DotNetCoreRestore(solutionFullPath);
+        var settings = new DotNetCoreRestoreSettings
+        {
+          Verbosity = Minimal
```
Nit: we indent 4 spaces.

durran replied:
Done

rstam commented on build.cake:
```diff
@@ -148,6 +154,7 @@ Task("Test")
             case "testnet452": settings.Framework = "net452"; break;
             case "testnetstandard15": settings.Framework = "netcoreapp1.1"; break;
             case "testnetstandard20": settings.Framework = "netcoreapp2.1"; break;
+            case "testnetstandard21": settings.Framework = "netcoreapp3.1"; break;
```
So this is confusing becase currently the `case` is the framework the build artifacts are targeting, but the `settings.Framework` is the framework that will be used to *run* the tests.

The framework we run the *tests* on ends up indirectly determining which build artifact is selected (i.e. which target framework the build artifact targets).

Since we don't target `netstandard21` I would not use that in a `case`. I would use `testnetcoreapp3.1`.

I would not mind if you changed the other three values to also use the framework to *test* on instead of the framework the build artifact *targets* (or maybe even allow both where there is an overlap?).

rstam replied:
Note: whether we run the tests using .NET Core 2.1 or .NET Core 3.1 in either case the artifacts that will be selected will be the ones that target .NET Standard 2.0.

So we are testing the same artifacts twice using different versions of .NET Core.

Nothing wrong with that... just making sure it's clear what is happening.

durran replied:
Changed to testnetcoreapp3.1

rstam commented on build.ps1:
```diff
@@ -5,11 +5,14 @@
 ##########################################################################
 
 <#
+
```
This file usually comes verbatim from the Cake repo.

If you made any local changes they will be lost next time an updated copy of this file is fetched from the Cake repo.

I can't tell if you made local changes or not.

durran replied:
I pulled this file directly from Cake but I did need to add one line to make it work with Cake as a tool. When we finally drop support for .NET Standard 1.5 we can just use the file from Cake directly with no modifications. https://github.com/durran/mongo-csharp-driver/pull/2/files#diff-3258bc4162690adead72c70b672093e68c92e86a4628148f0015d61738f21b5aR255

DmitryLukyanov replied:
Have you used this file: https://github.com/cake-build/resources/blob/master/dotnet-framework/build.ps1?
If so, I see some more small changes in the original file, they don't look too serious, is there a reason why they missed?

durran replied:
The diff that was highlighted I'm not sure we needed. We define the $Script value as build.cake on line 40, so I don't know why we would need to check if it exists later and set it. If we simply want the exact same one as the link you sent I can use it verbatim, since it's slightly different.

rstam replied:
It's still not clear to me whether you modified this file locally after fetching the latest copy from the Cake repo.

Ideally we would NOT modify this file. If you DID modify it how will we make sure such modifications are preserved in the future when someone overwrites this file with a newer version from the Cake repo?

durran replied:
I pulled this file directly from the Cake repo - the only thing I modified is where this comment is. https://github.com/durran/mongo-csharp-driver/pull/2/files/a9f97448fe0aee826bed2a3ca9f443807e04c978#r516939866

rstam replied:
OK. But:

> how will we make sure such modifications are preserved in the future when someone overwrites this file with a newer version from the Cake repo?

rstam commented on build.sh:
```diff
@@ -1,101 +1,77 @@
-##########################################################################
-
-# Define directories.
+# Define varibles
```
This file usually comes verbatim from the Cake repo.

If you made any local changes they will be lost next time an updated copy of this file is fetched from the Cake repo.

I can't tell if you made local changes or not.

DmitryLukyanov replied:
can you provide a link to the file which you used as an example here? You probably used some other link than this: https://github.com/cake-build/resources/blob/master/dotnet-framework/build.sh?

durran replied:
The diff I made was to make sure we had multiple .NET cores installed if they didn't exist, since the Cake build scripts assume that only 1 version of .NET core would be installed and no more. Now with https://jira.mongodb.org/browse/BUILD-12367 finished this is no longer needed and is removed.

durran replied:
Actually without this change on my latest run it does fail, so I think these are still needed.

Error is `The framework 'Microsoft.NETCore.App', version '2.1.0' was not found.`

rstam replied:
It's still not clear to me whether you modified this file locally after fetching the latest copy from the Cake repo.

Ideally we would NOT modify this file. If you DID modify it how will we make sure such modifications are preserved in the future when someone overwrites this file with a newer version from the Cake repo?

durran replied:
This file also comes directly from the Cake repo. The only thing I modified in this case is in the comment here: https://github.com/durran/mongo-csharp-driver/pull/2/files/a9f97448fe0aee826bed2a3ca9f443807e04c978#r523392217

rstam reviewed (changes requested):
Not so much changes requested as questions.

rstam commented:
> Not so much changes requested as questions.

Actually there were some changes requested, but also questions.

durran commented on build.cake:
```diff
@@ -37,6 +40,20 @@ var srcProjectNames = new[]
     "MongoDB.Driver.GridFS"
 };
 
+public static bool IsSuiteRunnable(string framework)
```
OCSP was only supported starting in .NET Core 3.0 on Linux, so we don't run the tests if we're testing OCSP on a lower version on Linux

rstam commented:
>> Open question. I used compile time directives to skip tests on Linux as we had a similar pattern with .NET standard version branches. I do feel, however, that maybe creating a custom attribute is better in most cases, like SkippableIfNotWindows or the like. Thoughts? Or maybe make it a separate ticket?

> I would say the current way is more flexible since we can use it in the code (not only in the tests) in future

Skipping tests is good in the sense that they get reported as skipped. If you #ifdef them away it's as if they never even existed.

But if you do decide to skip tests (instead of #ifdefing them away) we should follow the existing pattern for other test requiremenst and do something like:

    RequirePlatform.Check().Platform(...)

rstam commented:
Note: I recommend not force pushing to a PR because github gets confused and loses comments when you do that.

We usually wait until the PR is approved before squashing and then rebasing the branch on master.

Sometimes you have to force push... like when you need to rebase your branch on a newer version of master in order to get some other changes needed to continue work on the branch you are working on.

rstam commented on build.cake:
```diff
@@ -148,16 +172,21 @@ Task("Test")
             case "testnetstandard20": settings.Framework = "netcoreapp2.1"; break;
+            case "testnetcoreapp3.1": settings.Framework = "netcoreapp3.1"; break;
+        }
+        if (IsSuiteRunnable(settings.Framework))
```
I don't understand this check.

You are skipping the ENTIRE test suite just because an OCSP environment variable exists?

Shouldn't we be running the test suite and just skipping as necessary the individual tests that are OCSP related?

durran replied:
From my understanding there are not OCSP specific tests. It just changes the connection string, adds a few environment variables, changes the running cluster, and then runs the entire test suite. So as you can see in the current failures for OCSP all tests are failing that actually hit the database with auth errors.

durran commented:
The force pushing is because rebasing against master became quite a problem with hundreds of commits on this branch. It's much easier for me to rebase if I'm constantly squashing into a single commit. For future work I'll make sure to keep the PRs small so that's not an issue.

DmitryLukyanov commented:
@durran, yesterday, we discussed FLE for Linux, and in the scope of this discussion, we mentioned your PR. It looks like in the final version of your PR, we need to have Linux EG patches in a similar view as we have now for windows I mean that each target framework should be in their own test run.
![image](https://user-images.githubusercontent.com/44870738/99080792-d7dee100-25d2-11eb-85f3-e47f957d3b86.png)

durran commented on build.sh:
```diff
@@ -1,101 +1,77 @@
+      mkdir "$SCRIPT_DIR/.dotnet"
     fi
+    curl -Lsfo "$SCRIPT_DIR/.dotnet/dotnet-install.sh" https://dot.net/v1/dotnet-install.sh
+    bash "$SCRIPT_DIR/.dotnet/dotnet-install.sh" --channel 1.1 --install-dir .dotnet --no-path
```
This is a slight deviation from the Cake version in that we test across multiple .NET Core runtimes. The common pattern is to test against the minimum for the project but we need at least 3 installed - 1.1, 2.1, and 3.1 in this case.

durran commented on build.sh:
```diff
@@ -1,101 +1,77 @@
-    exec mono "$CAKE_EXE" $SCRIPT -verbosity=$VERBOSITY -configuration=$CONFIGURATION -target=$TARGET $DRYRUN "${SCRIPT_ARGUMENTS[@]}"
-fi
\ No newline at end of file
+(exec "$CAKE_EXE" build.cake --bootstrap) && (exec "$CAKE_EXE" build.cake "$@")
```
We don't want to require mono like Cake did previously and use the modern pattern of using Cake as a .NET tool cross platform. Need to bootstrap first to allow for the `dotnet` directive to work in the `build.cake`.

durran commented on evergreen/add-certs-if-needed.sh:
```diff
@@ -17,9 +17,15 @@ if [[ "$SSL" != "ssl" ]]; then
     certutil.exe -addstore "Root" ${DRIVERS_TOOLS}/.evergreen/ocsp/${OCSP_ALGORITHM}/ca.pem
   fi
+else
+  keytool -import -trustcacerts -file ${DRIVERS_TOOLS}/.evergreen/x509gen/ca.pem -keystore Root -storepass changeit -noprompt -alias x509
```
Adding certs on Linux via keytool for OCSP.

durran commented on evergreen/prepare-ocsp.sh:
```diff
@@ -0,0 +1,18 @@
+  /cygdrive/c/python/python38/python.exe -m venv ./venv
+  ./venv/Scripts/pip3 install -r ${DRIVERS_TOOLS}/.evergreen/ocsp/mock-ocsp-responder-requirements.txt
+else
+  which python
```
The OCSP tests on master fail here as well, this is just a start to get moving forward when that ticket lands.

DmitryLukyanov replied:
Sorry, not sure that I understand your comment, do you say that this script is not finished yet? If so, I would comment it out with comments about the current state and expected further steps

durran replied:
Added comment.

durran commented on tests/MongoDB.Driver.Core.Tests/Core/Operations/GeoSearchOperationTests.cs:
```diff
@@ -176,7 +176,7 @@ public void CreateCommand_should_return_expected_result_when_MaxTime_is_set(long
             bool async)
         {
-            RequireServer.Check().ClusterTypes(ClusterType.Standalone, ClusterType.ReplicaSet);
+            RequireServer.Check().ClusterTypes(ClusterType.Standalone, ClusterType.ReplicaSet).VersionLessThan("4.8.0");
```
Haystack indexes removed for next server release.

DmitryLukyanov commented on .gitignore:
```diff
@@ -66,3 +66,4 @@ src/packages
 artifacts
 packages
 /Tools
+.config/
```
Can you describe what the target for this?

durran replied:
I removed this change - nothing is outputing to .config anymore.

DmitryLukyanov commented on Release Notes/Release Notes v2.11.4.md:
```diff
@@ -0,0 +1,19 @@
+# .NET Driver Version 2.11.4 Release Notes
```
I think you need to create a new PR, since now we have too many unrelated changes

DmitryLukyanov commented on build.cmd:
```diff
@@ -1,5 +0,0 @@
-@echo off
```
Can you describe this change a bit?

durran replied:
build.cmd was removed since it is not used in any environment.

rstam replied:
Removing this file should be done as part of a different ticket.

See: https://jira.mongodb.org/browse/CSHARP-3246

durran replied:
Brought back in favour of the other ticket.

rstam commented:
Did you say you were going to make a new PR because the force push messed this PR up?

durran commented:
> Did you say you were going to make a new PR because the force push messed this PR up?

I got it fixed on this PR. All the extra commits are gone now.

DmitryLukyanov commented on build.cake:
```diff
@@ -141,6 +149,8 @@ Task("Test")
             NoRestore = true,
             Configuration = configuration,
+            Logger = "trx",
+            ResultsDirectory = outputDirectory.Combine("test-results"),
```
this looks related to CSHARP-2661

DmitryLukyanov replied:
also, not sure that `test-results` should be in `build` folder. I would say it should be either a new folder in `tests` folder or in `artifacts` where we place all script results like `hugo` or `nuget` files

durran replied:
I've removed this output now and will change in the other PR.

DmitryLukyanov commented on evergreen/evergreen.yml:
```diff
@@ -14,7 +14,7 @@ command_type: system
 # Protect ourself against rogue test case, or curl gone wild, that runs forever
 # 12 minutes is the longest we'll ever run
-exec_timeout_secs: 1800
+exec_timeout_secs: 4800
```
this change is unnecessary as far as I understand, now each TF task is launched in parallel

DmitryLukyanov commented on evergreen/evergreen.yml:
```diff
@@ -48,7 +48,7 @@ functions:
 
            # Python has cygwin path problems on Windows. Detect prospective mongo-orchestration home directory
-           if [ "Windows_NT" == "$OS" ]; then # Magic variable in cygwin
+           if [ "$OSTYPE" = "cygwin" ]; then # Magic variable in cygwin
```
can you describe this change a bit?

durran replied:
From what I recall with Evergreen the more consistent way of determining if we are on cygwin was checking the default $OSTYPE environment variable.

DmitryLukyanov replied:
Does `cygwin` always mean the same thing as Windows? I would say that the condition with `Windows_NT` is clearer says about what logic is actually expected here. How the same condition will look like on macOS?

durran replied:
I don't know why the github UI is showing this - I reverted the change back. Maybe because a comment is on it?

durran replied:
<img width="813" alt="Screenshot 2020-12-07 at 17 19 21" src="https://user-images.githubusercontent.com/9030/101376034-5e27d380-38b0-11eb-816c-0d2d40cd8e88.png">

DmitryLukyanov commented on evergreen/evergreen.yml:
```diff
@@ -1174,23 +1202,53 @@ buildvariants:
+     - name: test-netstandard15
+     - name: test-netstandard20
+
+#- matrix_name: "tests-snappy-compression-posix"
```
works only on `netstdard2.1`?

durran replied:
I've removed these for now, won't work in general on Linux at the moment.

DmitryLukyanov commented on evergreen/evergreen.yml:
```diff
@@ -661,6 +667,13 @@ tasks:
           vars:
             FRAMEWORK: netstandard20
 
+    - name: test-netstandard21
```
if we don't use it now, let's comment it out

durran replied:
Removed.

DmitryLukyanov commented on evergreen/evergreen.yml:
```diff
@@ -268,6 +268,7 @@ functions:
             FRAMEWORK=${FRAMEWORK} \
             evergreen/run-tests.sh
           echo "Skipping certificate removal..."
+          ls -la build/test-results
```
not sure whether it's required or not, but it looks harmless anyway

JamesKovacs replied:
This looks like debugging code trying to get test results to display in Evergreen. If so, we should revert these changes.

durran replied:
Reverted here.

DmitryLukyanov commented on tests/MongoDB.Driver.Legacy.VB.Tests/MongoDB.Driver.Legacy.VB.Tests.vbproj:
```diff
@@ -173,4 +173,4 @@
   -->
-</Project>
\ No newline at end of file
+</Project>
```
can you please revert this, it looks like nothing valuable

durran replied:
Revert the exclusion? These tests don't work on Linux and we discussed in a standup that we could remove from the project as a first pass, but leave the tests there and not delete them for the time being.

DmitryLukyanov replied:
I think we agreed to remove this diff?

DmitryLukyanov replied:
I meant that the whole diff for this file should not exist since no real changes here :)

DmitryLukyanov commented on tests/MongoDB.Driver.Tests/EncryptionTests.cs:
```diff
@@ -271,4 +272,6 @@ public static IMongoClient _mongocryptdClient(this AutoEncryptionLibMongoCryptCo
-}
+#endif
+    }
+}
```
as far as I know, we place a new empty line at the end of each file. That's why this red circle is presented here.

DmitryLukyanov replied:
I think we discussed this diff offline. cc: @durran

DmitryLukyanov commented on tests/MongoDB.Bson.Tests/Serialization/Serializers/CollectionSerializerGenericTests.cs:
```diff
@@ -276,6 +276,8 @@ public void TestTwoStrings()
     }
 
+#if Windows
+
```
empty line

DmitryLukyanov commented on tests/MongoDB.Bson.Tests/Serialization/Serializers/CollectionSerializerGenericTests.cs:
```diff
@@ -355,6 +357,8 @@ public void TestOneInt()
         }
     }
 
```
unnecessary empty line

DmitryLukyanov commented on evergreen/compile.sh:
```diff
@@ -9,5 +9,14 @@ set -o errexit  # Exit the script with error if any of the commands fail
+  powershell.exe .\\build.ps1 -target Build -Verbosity Diagnostic
+else
+  for var in TMP TEMP NUGET_PACKAGES NUGET_HTTP_CACHE_PATH APPDATA; do
+    export $var=/data/tmp;
```
nit: it will be good to have all `if` code paths as more similar as possible to highlight that the only difference between them is system paths

JamesKovacs replied:
Am I understanding the logic correctly that we only compile on Windows under cygwin? For Linux, we just use pre-built nuget packages? If so, I would suggest we echo "Skipping" or something to that effect in the else clause to avoid future confusion. Right now our output logs on Linux are going to say "Compiling .NET driver" with no indication that it isn't actually compiled on that platform.

durran replied:
Actually you caught a bug. I've updated it to actually compile both.

rstam commented on evergreen/evergreen.yml:
```diff
@@ -14,7 +14,7 @@ command_type: system
 # Protect ourself against rogue test case, or curl gone wild, that runs forever
 # 12 minutes is the longest we'll ever run
-exec_timeout_secs: 1800
+exec_timeout_secs: 4800
```
80 minutes?

Is this change justified?

durran replied:
I have changed back. The Linux builds were taking too long because of another error that I resolved, in which all target test variants were running on every build.

rstam commented on evergreen/evergreen.yml:
```diff
@@ -48,7 +48,7 @@ functions:
 
            # Python has cygwin path problems on Windows. Detect prospective mongo-orchestration home directory
-           if [ "Windows_NT" == "$OS" ]; then # Magic variable in cygwin
+           if [ "$OSTYPE" = "cygwin" ]; then # Magic variable in cygwin
```
cygwin sucks. And it's only a Windows thing right?

Why are we using it?

JamesKovacs replied:
Cygwin provides a bash environment on Windows with all the assorted Unix-y tools. I suspect that we are using it so we can leverage existing Evergreen bash scripts. Regardless of the reason, it is outside the scope of this PR to switch scripting languages IMHO. (And yes, Cygwin has some rough edges. It is 'nixy enough to make you think you're using Unix most of the time but you end up stubbing your toes frequently.)

durran replied:
Yeah I would love to switch the build scripts on Windows over to Powershell only but that's a pretty big effort from what I can see at the moment.

rstam commented on evergreen/evergreen.yml:
```diff
@@ -195,7 +195,7 @@ functions:
     - command: attach.xunit_results
       params:
-        file: ./mongo-csharp-driver/*/build/test-results/TEST-*.xml
+        file: mongo-csharp-driver/build/test-results/*.trx
```
Is this actually related to CSHARP-1847?

JamesKovacs replied:
Looks like it is related to CSHARP-2661. We should keep the PRs separate. And if we need CSHARP-2661 fixed before we can commit CSHARP-1847, then we should review and pull CSHARP-2661 onto master first. Then we can rebase CSHARP-1847 onto it.

durran replied:
I've reverted this here.

rstam commented on evergreen/evergreen.yml:
```diff
@@ -288,6 +289,15 @@ functions:
           ${PREPARE_SHELL}
           MONGODB_URI="${plain_auth_mongodb_uri}" evergreen/run-plain-auth-tests.sh
 
+  run-gssapi-auth-tests:
```
Is this actually related to CSHARP-1847?

DmitryLukyanov replied:
who set this value: `gssapi_mongodb_uri`? I would expect something like this: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/84/files#diff-385f26e0b474adecd1115bc31e81b481a19a922a2c7096b444f1a998d233f157R239. NOTE: values like `AUTH_HOST` and `AUTH_GSSAPI` are already defined on the EG

DmitryLukyanov replied:
Maybe it's better to work on this change in the scope of a different ticket?

DmitryLukyanov replied:
NOTE: the value which is defined on EG now is:
![image](https://user-images.githubusercontent.com/44870738/99705869-5e9c2e00-2aab-11eb-8051-98e37e07db4e.png)
which is quite different from the one mentioned in the script. Is it a new value? I would say it's not relly good to have different options for the same value

durran replied:
I've reverted these changes here.

rstam commented on evergreen/evergreen.yml:
```diff
@@ -472,9 +482,8 @@ functions:
-          /cygdrive/c/python/python38/python.exe -m venv ./venv
-          ./venv/Scripts/pip3 install -r ${DRIVERS_TOOLS}/.evergreen/ocsp/mock-ocsp-responder-requirements.txt
+          ${PREPARE_SHELL}
+          evergreen/prepare-ocsp.sh
```
I have no idea why this change was made. (As an aside driver-tools is a nightmare to understand and use).

DmitryLukyanov replied:
as far as I can see, this logic has been moved into `prepare-ocsp.sh` with adding additional logc

durran replied:
I was trying to get the OCSP tests working and refactored out common logic before I realised they wouldn't pass on .NET Core < 3.0.

rstam commented on build.cake:
```diff
@@ -141,6 +149,8 @@ Task("Test")
             NoBuild = true,
             NoRestore = true,
             Configuration = configuration,
+            Logger = "trx",
```
Is this change actually related to CSHARP-1847?

durran replied:
Removed.

rstam commented on build.config:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+CAKE_VERSION="0.38.4"
+DOTNET_VERSION="3.1.403"
```
Is this a shell script? Shouldn't the extension be `.sh`?

durran replied:
This is just a file read by build.sh and build.ps1

DmitryLukyanov commented on evergreen/evergreen.yml:
```diff
@@ -510,9 +518,8 @@ functions:
       params:
         script: |
-          cd ${DRIVERS_TOOLS}/.evergreen/ocsp
-          /cygdrive/c/python38/python.exe -m venv ./venv
```
was it a bug here in the path (and below)?

compare:
> /cygdrive/c/python38/python.exe

and

>/cygdrive/c/python/python38/python.exe

durran replied:
They all use `/cygdrive/c/python/python38/python.exe` now from prepare-ocsp.sh

DmitryLukyanov replied:
yep, I just wanted to highlight that we probably had a bug here. So, not sure why it didn't affect our tests.. cc: @rstam , @JamesKovacs

DmitryLukyanov commented on build.cake:
```diff
@@ -49,7 +53,11 @@ Task("Release")
     .Does(() =>
     {
-        DotNetCoreRestore(solutionFullPath);
+        var settings = new DotNetCoreRestoreSettings
```
nit: do we really need this change?

durran replied:
Removed.

DmitryLukyanov commented on evergreen/run-gssapi-tests.sh:
```diff
@@ -0,0 +1,21 @@
+    exit 1
+fi
+export MONGODB_URI="${MONGODB_URI}&authSource=\$external"
+powershell.exe \
```
not sure how it works. The connection string which I used in my case was built in the c# code and looked like:

     return $"mongodb://{authGssapi}@{authHost}/kerberos?authMechanism=GSSAPI{mechanismProperty}";

as far as I can see, you don't set these values. 
NOTE: the mentioned variables are already defined on EG

DmitryLukyanov replied:
do we actually run it? I don't see it here https://evergreen.mongodb.com/version/5fb67d9c3627e066037f846b

durran replied:
I took out running it as it's part of another ticket.

DmitryLukyanov replied:
so, it should be removed now :) ?

durran replied:
Removed.

JamesKovacs commented on evergreen/run-tests.sh:
```diff
@@ -102,26 +102,45 @@ if [[ "$OS" =~ Windows|windows ]]; then
+    export $var=z:\\data\\tmp;
+  else
+    export $var=/data/tmp;
+  fi
```
This pattern repeats often. We should define a global TMP_PATH variable that we set initially based on the OS and then use $TMP_PATH as needed.

DmitryLukyanov replied:
I agree with @JamesKovacs in general, but maybe if we don't do it, it will be easier for us since the diff will be clearer, just thinking out loud?

JamesKovacs replied:
Sometimes we use `if [ COND ]` and other times `if [[ COND ]]`. The latter is more modern and preferred as it provides additional features. We should standardize on the latter throughout our Bash scripts.

jyemin replied:
Can we defer this change? It seems stylistic.

DmitryLukyanov commented:
@durran , I confirmed with John Chen, that the solution for this ticket https://jira.mongodb.org/browse/BUILD-12367 will include `.netcoreapp1.1`.
Can you please check the made changes and find which of the scripts (if any) will be no longer required when the build ticket will be done? (as I understand it should be ready tomorrow or so)

DmitryLukyanov commented on evergreen/compile.sh:
```diff
@@ -9,5 +9,15 @@ set -o errexit  # Exit the script with error if any of the commands fail
+else
+  for var in TMP TEMP NUGET_PACKAGES NUGET_HTTP_CACHE_PATH APPDATA; do
+    export $var=/data/tmp;
+    ./build.sh -target=Build -verbosity=Diagnostic
```
is it expected that for a Cygwin branch, we call PowerShell outside the `for` brackets, but in the `else` code path, we call scripts inside the `for`?

DmitryLukyanov replied:
As I mentioned above, it would be good to have the same branches with the difference in the used variables, at least if it's possible

durran replied:
The build.sh getting executed in the for was a bug, I've move it out of the for loop.

DmitryLukyanov commented on evergreen/run-plain-auth-tests.sh:
```diff
@@ -11,3 +11,11 @@ set -o errexit  # Exit the script with error if any of the commands fail
 ############################################
 
 echo "Running PLAIN authentication tests"
+if [ -z ${MONGODB_URI+x} ]; then
```
should be removed?

durran replied:
Done

DmitryLukyanov commented on tests/MongoDB.Bson.Tests/Jira/CSharp783Tests.cs:
```diff
@@ -142,6 +144,8 @@ public void TestSortedSetTwoInts(int x, int y)
         }
     }
 
+#endif
```
I think we don't have empty lines between `ifdef` conditions and their inner content

durran replied:
Done

DmitryLukyanov commented on tests/MongoDB.Bson.Tests/MongoDB.Bson.Tests.csproj:
```diff
@@ -17,13 +17,25 @@
+  </PropertyGroup>
+  
+  <PropertyGroup Condition="'$(IsWindows)'=='true'">
+    <DefineConstants>WINDOWS</DefineConstants>
```
Just to highlight: as far as I understand, this code will overwrite line 29 (with Trace). So this value won't be available anymore. It looks like we don't use it at all, so it looks safe, but if we agree with it, I think we need to remove lines 28-30.
cc: @rstam , @JamesKovacs

DmitryLukyanov replied:
similar in other places

durran replied:
Instead of removing those lines I've just made sure the constants are not overridden. Like so: `<DefineConstants>$(DefineConstants);WINDOWS</DefineConstants>`

DmitryLukyanov commented on tests/MongoDB.Driver.Tests/EncryptionTests.cs:
```diff
@@ -29,11 +29,12 @@
     public class EncryptionTests
     {
-        #region static
+    #region static
```
this diff (and line 37) is wrong and should be reverted

durran replied:
Done

DmitryLukyanov commented on tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/ClientSideEncryptionTestRunner.cs:
```diff
@@ -316,13 +318,13 @@ private void ReplaceTypeAssertionWithActual(BsonArray actual, BsonArray expected
     public class TestCaseFactory : JsonDrivenTestCaseFactory
     {
-        #region static
+#region static
```
this diff (and line 327) should be reverted

durran replied:
Done

DmitryLukyanov commented on tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:
```diff
@@ -1121,7 +1122,7 @@ private IMongoCollection<BsonDocument> GetCollection(IMongoClient client, Collec
         public class JsonFileReader : EmbeddedResourceJsonFileReader
         {
-            #region static
+#region static
```
should be reverted

durran replied:
Done

DmitryLukyanov commented on tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs:
```diff
@@ -1131,7 +1132,7 @@ public class JsonFileReader : EmbeddedResourceJsonFileReader
             // public static properties
             public static JsonFileReader Instance => __instance.Value;
-            #endregion
+#endregion
```
should be reverted

durran replied:
Done

DmitryLukyanov commented on tests/MongoDB.Bson.Tests/Jira/CSharp783Tests.cs:
```diff
@@ -23,6 +23,8 @@
 
 namespace MongoDB.Bson.Tests.Jira.CSharp783
 {
+
```
unnecessary empty line

durran replied:
Done

durran replied:
Now pushed. :)

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Core/Compression/ZstandardNativeTests.cs:
```diff
@@ -28,7 +28,8 @@ namespace MongoDB.Driver.Core.Tests.Core.Compression
     {
-        #region static
+#if WINDOWS
+#region static
```
#region lines should be reverted

durran replied:
Done

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs:
```diff
@@ -92,9 +92,10 @@ public void RunTestDefinition(JsonDrivenTestCase testCase)
                 }
             }
-
+#if WINDOWS
```
we should create a ticket to restore it on linux

durran replied:
https://jira.mongodb.org/browse/CSHARP-3274

DmitryLukyanov commented on evergreen/evergreen.yml:
```diff
@@ -220,10 +220,10 @@ functions:
+          AUTH=${AUTH} \
+          SSL=${SSL} \
+          ORCHESTRATION_FILE=${ORCHESTRATION_FILE} \
             sh ${DRIVERS_TOOLS}/.evergreen/run-orchestration.sh
```
are the spaces before `sh` line should also be adjusted?

durran replied:
I was just having it match the formatting of the bootstrap-mongo-orchestration function

DmitryLukyanov commented:
I think we need to consider logic like this: https://github.com/mongodb/libmongocrypt/blob/master/bindings/cs/MongoDB.Libmongocrypt/MongoDB.Libmongocrypt.csproj#L4-L5. Otherwise, the `dotnet test` command will fail if we run it without direct specifying the target framework on Linux since Linux doesn't know about `net452`

durran commented:
@DmitryLukyanov I used Cake's actual scripts that it uses itself: https://github.com/cake-build/cake/blob/develop/build.ps1 and https://github.com/cake-build/cake/blob/develop/build.sh since https://github.com/cake-build/resources/tree/master/dotnet-core all have no logic in them

DmitryLukyanov commented on evergreen/evergreen.yml:
```diff
@@ -48,7 +48,8 @@ functions:
            # Python has cygwin path problems on Windows. Detect prospective mongo-orchestration home directory
-           if [ "Windows_NT" == "$OS" ]; then # Magic variable in cygwin
+           if [ "Windows_NT" = "$OS" ]; then # Magic variable in cygwin
+              echo "########################## CYGWIN"
```
nit: this may be reworded :)

durran replied:
I removed the echo - was just a test for evergreen.

DmitryLukyanov commented on evergreen/evergreen.yml:
```diff
@@ -279,15 +280,6 @@ functions:
           # DO NOT ECHO WITH XTRACE (which PREPARE_SHELL does)
           ATLAS_FREE="${ATLAS_FREE}" ATLAS_FREE_SRV="${ATLAS_FREE_SRV}" ATLAS_REPLICA="${ATLAS_REPLICA}" ATLAS_REPLICA_SRV="${ATLAS_REPLICA_SRV}" ATLAS_SHARDED="${ATLAS_SHARDED}" ATLAS_SHARDED_SRV="${ATLAS_SHARDED_SRV}" ATLAS_TLS11="${ATLAS_TLS11}" ATLAS_TLS11_SRV="${ATLAS_TLS11_SRV}" ATLAS_TLS12="${ATLAS_TLS12}" ATLAS_TLS12_SRV="${ATLAS_TLS12_SRV}" evergreen/run-atlas-connectivity-tests.sh
 
-  run-plain-auth-tests:
```
nit: this (and related changes) may be reverted to make diff smaller

rstam replied:
Why are these tests being removed?

durran replied:
Done

DmitryLukyanov reviewed (commented):
> 
> 
> @DmitryLukyanov I used Cake's actual scripts that it uses itself: https://github.com/cake-build/cake/blob/develop/build.ps1 and https://github.com/cake-build/cake/blob/develop/build.sh since https://github.com/cake-build/resources/tree/master/dotnet-core all have no logic in them

Thanks. Can you please also check:
- build.sh:
![image](https://user-images.githubusercontent.com/44870738/101039258-daf33e80-358c-11eb-89fb-daf078daf795.png)
is the selected diff required?

- build.ps1:
as I understand it was taken from here: https://github.com/cake-build/resources/blob/master/dotnet-framework/build.ps1 because the diff again the link you mentioned is too big. If so, are these diffs required:
![image](https://user-images.githubusercontent.com/44870738/101041719-8b614280-358d-11eb-889b-a64eea998d6b.png)

DmitryLukyanov commented:
When the latest comments and this https://github.com/durran/mongo-csharp-driver/pull/2#discussion_r530972973 will be resolved, this will be LGTM

JamesKovacs commented on evergreen/prepare-ocsp.sh:
```diff
@@ -0,0 +1,17 @@
+  # Need to ensure on Linux python is installed in the correct place and visible to the script.
+  # /opt/python/2.7/bin/python -m venv ./venv
+  # ./venv/Scripts/pip3 install -r ${DRIVERS_TOOLS}/.evergreen/ocsp/mock-ocsp-responder-requirements.txt
+fi
```
Do we have a separate CSHARP ticket to get OCSP tests on Linux working? Right now the non-cygwin branch to install dependencies is just commented out.

durran replied:
https://jira.mongodb.org/browse/CSHARP-3255

JamesKovacs replied:
Then we should note the ticket and reason in a comment.

durran replied:
Added code comment

JamesKovacs replied:
Perfect. Thanks.

JamesKovacs commented on evergreen/run-tests.sh:
```diff
@@ -102,26 +102,45 @@ if [[ "$OS" =~ Windows|windows ]]; then
+    export TARGET="Test"
+  else
+    export TARGET="Test${FRAMEWORK}"
+  fi
```
If `$FRAMEWORK` is `nil`, then `"Test${FRAMEWORK}"` equals `Test`. Then again, explicitly calling out that `$FRAMEWORK` could be `nil` is more obvious. Up to you whether you'd prefer the above or the terser:
`export TARGET="Test${FRAMEWORK}"`

durran replied:
Changed per your suggestion.

JamesKovacs commented on src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj:
```diff
@@ -25,17 +31,15 @@
     <Message Text="IsWindows: '$(IsWindows)'" Importance="high" />
     <Message Text="IsOSX: '$(IsOSX)'" Importance="high" />
     <Message Text="IsLinux: '$(IsLinux)'" Importance="high" />
   </Target>
```
This seems like debug info from troubleshooting the build. Is there a reason to still muddy the build output by dumping out these variables?

DmitryLukyanov replied:
This is information logs, Robert asked a similar question here: https://mongodb.slack.com/archives/GQ27C9N7J/p1606854006080700. In two words, I think we can remove it, but at the same time, I think this still can be helpful from time to time. Since It's harmless, I would not do it, but I don't insist on it

durran replied:
Removed, was a remnant of some debugging.

JamesKovacs commented on tests/MongoDB.Bson.Tests/Jira/CSharp275Tests.cs:
```diff
@@ -83,35 +83,36 @@ public Test(string json, string iso)
-            new Test("Mon, 10 Oct 2011 11:22:33 -0000", "2011-10-10T11:22:33-00:00"),
-            new Test("Mon, 10 Oct 2011 11:22:33 +0100", "2011-10-10T11:22:33+01:00"),
-            new Test("Mon, 10 Oct 2011 11:22:33 -0100", "2011-10-10T11:22:33-01:00")
+            // TODO: Investigate Military Time Failures
```
Note CSHARP-3268 as the ticket to investigate these failures.

durran replied:
Added code comment with link.

JamesKovacs replied:
Perfect.

JamesKovacs reviewed (changes requested):
Overall looking good. Only a few minor issues left to address.

rstam commented on evergreen/evergreen.yml:
```diff
@@ -665,10 +652,6 @@ tasks:
-    - name: plain-auth-tests
-      commands:
-        - func: run-plain-auth-tests
-
```
Why are these tests being removed?

durran replied:
Brought back now.

rstam commented on evergreen/evergreen.yml:
```diff
@@ -1136,9 +1124,15 @@ buildvariants:
     - windows-64-vs2017-compile
   tasks:
     - name: compile
+- name: ubuntu1804-test
```
Shouldn't this be named `ubuntu1804-compile`?

durran replied:
Fixed.

rstam commented on evergreen/evergreen.yml:
```diff
@@ -1211,13 +1221,6 @@ buildvariants:
-#     - windows-64-vs2017-test
-#   tasks:
-#     - name: plain-auth-tests
-
```
Instead of removing the commented lines maybe we should figure out why these tests aren't being run?

durran replied:
Brought back. Will be resolved in CSHARP-1849

rstam commented on evergreen/run-tests.sh:
```diff
@@ -92,36 +92,47 @@ fi
   fi
 else
-  export TARGET="Test"
+  export TARGET="Test${FRAMEWORK}"
```
Both branches of the `if` run the same `export` command.

Should the common line be moved to before the if?

durran replied:
Moved up.

rstam commented on src/MongoDB.Driver.Core/MongoDB.Driver.Core.csproj:
```diff
@@ -1,7 +1,13 @@
+    <IsWindows Condition="$([MSBuild]::IsOSPlatform('Windows'))">true</IsWindows>
+    <IsOSX Condition="$([MSBuild]::IsOSPlatform('OSX'))">true</IsOSX>
+    <IsLinux Condition="$([MSBuild]::IsOSPlatform('Linux'))">true</IsLinux>
+  </PropertyGroup>
```
I find having a bunch of `Is` properties confusing, why not:

    <PropertyGroup>
      <Platform Condition="$([MSBuild]::IsOSPlatform('Windows'))">Windows</IsWindows>
      <Platform Condition="$([MSBuild]::IsOSPlatform('OSX'))">OSX</IsOSX>
      <Platform Condition="$([MSBuild]::IsOSPlatform('Linux'))">Linux</IsLinux>
    </PropertyGroup>

Just a thought... what you have obviously works.

rstam replied:
One thing that really confused me about the family of `Is` properties is the use of the value `true`.

It *looks* like a boolean, but really is just a string (right?).

durran replied:
Yes it is actually a string in this case. I was just basically following the pattern set in https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/152/files

rstam commented on tests/MongoDB.Bson.Tests/Jira/CSharp783Tests.cs:
```diff
@@ -23,6 +23,7 @@
 
 namespace MongoDB.Bson.Tests.Jira.CSharp783
 {
+#if WINDOWS
```
I see that the preprocessor symbol `WINDOWS` *is* defined now (I don't believe it was before).

But the way it is defined seems fragile. I see that it is defined only in some `.csproj` files (presumably only the ones where it is needed). 

If we should happen to forget to define the `WINDOWS` preprocessor symbol this code will be silently commented out and we would be none the wiser.

Maybe this is the best we can do..., but it worries me.

I think we should have a JIRA ticket to **remove** all occurrences of `#if WINDOWS` in the future.

rstam replied:
Is there an existing preprocessor symbol we can use instead?

Maybe `_WIN64`?

rstam replied:
Also... from a previous comment I made that is no longer visible in the files view:

Skipping tests is good in the sense that they get reported as skipped. If you #ifdef them away it's as if they never even existed.

But if you do decide to skip tests (instead of #ifdefing them away) we should follow the existing pattern for other test requiremenst and do something like:

    RequirePlatform.Check().Platform(...)

rstam replied:
If you used `RequirePlatform` instead of `#if WINDOWS` we might not need the `WINDOWS` preprocessor symbol at all...

DmitryLukyanov replied:
I think we can use the solution from here: https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/163/files#diff-c05284cb429957908b8ffcd50b29c61eb26c0af4871e58699080d527b0ec2ffbR37

DmitryLukyanov replied:
I can work on this change in the scope of my ticket

DmitryLukyanov replied:
>
>Is there an existing preprocessor symbol we can use instead?
>Maybe _WIN64?

I never heard/seen this and wasn't able to google it now

durran commented:
@DmitryLukyanov Even with the build ticket to install all the .NET core versions on the Ubuntu machines finished, all versions of .NET core is still not found when running the tests unless I specifically install them with the script. I removed the lines doing that and the build failed so I added them back. I specifically added also the 1.1 and 2.1 channels since Cake's scripts make the assumption that the project is only testing against 1 version of .NET core and we test against more than 1. Hence the diff against the Cake scripts

DmitryLukyanov commented on evergreen/compile.sh:
```diff
@@ -9,5 +9,15 @@ set -o errexit  # Exit the script with error if any of the commands fail
 
-for var in TMP TEMP NUGET_PACKAGES NUGET_HTTP_CACHE_PATH APPDATA; do setx $var z:\\data\\tmp; export $var=z:\\data\\tmp; done
-powershell.exe .\\build.ps1 -target Build -Verbosity Diagnostic
+if [ "$OSTYPE" = "cygwin" ]; then
```
because of this comment https://github.com/durran/mongo-csharp-driver/pull/2#discussion_r526987531, should we use just `$OS`?

durran replied:
Done.

DmitryLukyanov commented on evergreen/prepare-ocsp.sh:
```diff
@@ -0,0 +1,17 @@
+
+echo "Preparing Python env for OCSP tests"
+
+if [ "$OSTYPE" = "cygwin" ]; then
```
because of this comment https://github.com/durran/mongo-csharp-driver/pull/2#discussion_r526987531, should we use `$OS`?

durran replied:
Done

DmitryLukyanov commented on evergreen/run-tests.sh:
```diff
@@ -92,36 +92,47 @@ fi
 fi
 for var in TMP TEMP NUGET_PACKAGES NUGET_HTTP_CACHE_PATH APPDATA; do
-  export $var=z:\\data\\tmp;
+  if [ "$OSTYPE" = "cygwin" ]; then
```
because of this comment https://github.com/durran/mongo-csharp-driver/pull/2#discussion_r526987531, should we use `$OS`?

durran replied:
Done

DmitryLukyanov reviewed (approved):
1. this is not done: https://github.com/durran/mongo-csharp-driver/pull/2/files#r537697733
2. Assuming that the comments from other reviewers are also resolved
3. Can you please post the latest EG link?

other than the above, LGTM, thanks great job!

JamesKovacs commented:
Two remaining points from me...

1. We have a few places where code is commented and a ticket exists to get it working. We should include the CSHARP ticket in the comments so we know why this code is commented.
- [add link to CSHARP-3255](https://github.com/durran/mongo-csharp-driver/pull/2#discussion_r538594901)
- [add link to CSHARP-3268](https://github.com/durran/mongo-csharp-driver/pull/2#discussion_r535478199)

2. Our bash scripts use a mix of if [ COND ] and if [[ COND ]]. It would be nice to standardize on the latter syntax, but might not be worth the effort at the present time.

I'd like to see the first resolved, but am ok with deferring the second. The rest LGTM.

durran commented:
@DmitryLukyanov Got your last change in - I had missed the line and did the one below it previously.

durran commented:
@JamesKovacs I've added links now to the two tickets in the original JIRA ticket's comments (https://jira.mongodb.org/browse/CSHARP-1847) - I'm assuming that's where you meant since they are in comments here already.

JamesKovacs reviewed (approved):
LGTM

rstam reviewed (approved):
LGTM

My only remaining concern is that the changes you have made to builds.ps1 and build.sh will be lost the next time we overwrite these files with the latest copies from the Cake repository.
